subsys/debug/tracing/include/tracing_backend.h

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2019 Intel corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef _TRACE_BACKEND_H
+#define _TRACE_BACKEND_H
+
+#include <sys/util.h>
+#include <tracing_packet.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Tracing backend
+ * @defgroup Tracing_backend Tracing backend
+ * @{
+ * @}
+ */
+
+#define TRACING_BACKEND_UART_NAME "tracing_backend_uart"
+
+struct tracing_backend;
+
+/**
+ * @brief Tracing backend API.
+ */
+struct tracing_backend_api {
+	void (*init)(void);
+	void (*output)(const struct tracing_backend *backend,
+			struct tracing_packet *packet);
+};
+
+/**
+ * @brief Tracing backend control block.
+ */
+struct tracing_backend_control_block {
+	void *ctx;
+};
+
+/**
+ * @brief Tracing backend structure.
+ */
+struct tracing_backend {
+	const char *name;
+	const struct tracing_backend_api *api;
+	struct tracing_backend_control_block *cb;
+};
+
+/**
+ * @brief Create tracing_backend instance.
+ *
+ * @param _name Instance name.
+ * @param _api  Tracing backend API.
+ */
+#define TRACING_BACKEND_DEFINE(_name, _api)                              \
+	static struct tracing_backend_control_block _name##_cb;          \
+	static const Z_STRUCT_SECTION_ITERABLE(tracing_backend, _name) = \
+	{                                                                \
+		.name = STRINGIFY(_name),                                \
+		.api = &_api,                                            \
+		.cb = &_name##_cb                                        \
```

Comment:
Would it be possible to avoid _pointing_ to the control block? It doesn't change as far as I can tell, so it could be inplace.